siphon/simplewebservice/igra2.py

```diff
@@ -0,0 +1,427 @@
+
+        body, header, dates_long, dates = self._select_date_range(lines)
+
+        return body, header, dates_long, dates
+
+    def _select_date_range(self, lines):
+        """Identify lines containing headers within the range begin_date to end_date.
+
+        Parameters
+        -----
+        lines: list
+            list of lines from the IGRA2 data file.
+        """
+        headers = []
+        num_lev = []
+        dates = []
+
+        # Get indices of headers, and make a list of dates and num_lev
+        for idx, line in enumerate(lines):
+            if line[0] == '#':
+                year, month, day, hour = map(int, line[13:26].split())
+
+                # All soundings have YMD, most have hour
+                try:
+                    date = datetime.datetime(year, month, day, hour)
+                except ValueError:
+                    date = datetime.datetime(year, month, day)
+
+                # Check date
+                if self.begin_date <= date <= self.end_date:
+                    headers.append(idx)
+                    num_lev.append(int(line[32:36]))
+                    dates.append(date)
+                if date > self.end_date:
+                    break
+
+        if len(dates) == 0:
+            # Break if no matched dates.
+            # Could improve this later by showing the date range for the station.
+            raise ValueError('No dates match selection.')
+
+        # Compress body of data into a string
+        begin_idx = min(headers)
+        end_idx = max(headers) + num_lev[-1]
+
+        # Make a boolean vector that selects only list indices within the time range
+        selector = np.zeros(len(lines), dtype=bool)
+        selector[begin_idx:end_idx+1] = True
+        selector[headers] = False
+        body = ''.join([line for line in itertools.compress(lines, selector)])
+
+        selector[begin_idx:end_idx + 1] = ~selector[begin_idx:end_idx + 1]
+        header = ''.join([line for line in itertools.compress(lines, selector)])
+
+        # expand date vector to match length of the body dataframe.
+        dates_long = np.repeat(dates, num_lev)
+
+        return body, header, dates_long, dates
+
+    def _get_fwf_params(self):
+        """Produce a dictionary with names, colspecs, and dtype for IGRA2 data.
+        
+        Returns a dict with entries 'body' and 'header'.
+        """
+
+        def _cdec(power=1):
+            """Make a function to convert string 'value*10^power' to float."""
+            def _cdec_power(val):
+                return float(val)/10**power
+            return _cdec_power
+
+        def _cflag(val):
+            """Replace alphabetic flags A and B with numeric."""
+            if val == 'A':
+                return 1
+            elif val == 'B':
+                return 2
+            else:
+                return 0
+
+        def _ctime(strformat='MMMSS'):
+            """Returns a function converting a string from MMMSS or HHMM to seconds.
```

Comment:
"Returns" -> "Return". Bots. Also go ahead and please move the closing `"""` on the next line up to this one.